nupic/research/frameworks/vernon/experiments/base_experiment.py

```diff
@@ -0,0 +1,184 @@
+        console.setFormatter(logging.Formatter(log_format))
+        self.logger = logging.getLogger(config.get("name", type(self).__name__))
+        self.logger.setLevel(log_level)
+        self.logger.addHandler(console)
+        self.progress = config.get("progress", False)
+        self.launch_time = config.get("launch_time", time.time())
+        self.logdir = config.get("logdir", None)
+
+        if config.get("disable_logger", False):
+            self.logger.disabled = True
+            self.progress = False
+
+        # Configure seed
+        self.seed = config.get("seed", 42)
+        set_random_seed(self.seed, False)
+
+        self.logger.info("Execution order: %s",
+                         pformat(self.get_execution_order()))
+
+    @abc.abstractmethod
+    def run_iteration(self):
+        """
+        Run one training iteration of the experiment and return some sort of
+        result.
+        """
+
+    @abc.abstractmethod
+    def should_stop(self):
+        """
+        Whether or not the experiment should stop. Usually determined by the
+        number of epochs but customizable to any other stopping criteria
+        """
+
+    def stop_experiment(self):
+        """
+        Perform any needed cleanup.
+        """
+
+    def run_pre_experiment(self):
+        """
+        Evaluate the model before performing any training and return some sort
+        of result. If the experiment is configured to skip the pre_experiment,
+        return None.
+        """
+        return None
+
+    @classmethod
+    def insert_pre_experiment_result(cls, result, pre_experiment_result):
+        """
+        Modify result to incorporate the pre_experiment_result. By default, this
+        method performs no update. (This may be sufficient, for example if the
+        pre_experiment_result has already been printed to the console.)
+
+        :param pre_experiment_results: The return value of pre_experiment
+        """
+
+    def get_state(self):
+        """
+        Get experiment serialized state as a dictionary of  byte arrays
+        :return: dictionary with "model", "optimizer" and "lr_scheduler" states
+        """
+        return {}
+
+    def set_state(self, state):
+        """
+        Restore the experiment from the state returned by `get_state`
+        :param state: dictionary with "model", "optimizer", "lr_scheduler"
+        states
+        """
+
+    @classmethod
+    def get_printable_result(cls, result):
+        """
+        Return a stripped down version of result that has its large data structures
+        removed so that the result can be printed to the console.
+        """
+        return result
+
+    @classmethod
+    def get_execution_order(cls):
+        """
+        Gets a dict that can be printed to show the order of events that occur
+        for each method. Subclasses and mixins should extend this method,
+        modifying each list of events according to when/whether they call
+        super().
```

Comment:
Generally, could you help me understand when something deserves to go into the execution order? Maybe we should describe it in the docstring here.

Reply:
Good point, I added it to the interface docstring.